pkg/agent/controller/networkpolicy/cache.go

```diff
@@ -91,10 +91,10 @@ func hashRule(r *rule) string {
 // It's the struct used by reconciler.
 type CompletedRule struct {
 	*rule
-	// Source Pods of this rule, can't coexist with ToAddresses.
-	FromAddresses v1beta1.GroupMemberPodSet
-	// Destination Pods of this rule, can't coexist with FromAddresses.
-	ToAddresses v1beta1.GroupMemberPodSet
+	// Source GroupMembers of this rule, can't coexist with ToAddresses.
+	FromAddresses v1beta1.GroupMemberSet
+	// Destination GroupMembers of this rule, can't coexist with FromAddresses.
+	ToAddresses v1beta1.GroupMemberSet
 	// Target Pods of this rule.
 	Pods v1beta1.GroupMemberPodSet
```

Comment:
Why target cannot be external endpoints?

Reply:
I believe the reason is that for the externalEntities in the appliedTo, there's nothing to be enforced from the Antrea agent side. The antrea-cloud controller will enforce firewall rules on the cloud provider instead.

Reply:
Ok. I thought we will support agented model together.

Reply:
would let @suwang48404 comment on that

Reply:
If we do not support agented model for external Nodes, then we need not any code in agent to handle ExternalEndpoints/Nodes (I think you still have such code?)?

Reply:
We need that code in case there are pods in the appliedTo and externalEntities in the ingress/egress rules.

Reply:
I think we should support ExternalEndpoints in from/to/appliedTo together; if no support for appliedTo, no need to support from/to either.

Reply:
this is my view too.. we should get the controller part ready as well in order to merge this PR.. that won't happen before tomorrows release, so i would like to aim at merging these PRs in the next release

Reply:
> I think we should support ExternalEndpoints in from/to/appliedTo together; if no support for appliedTo, no need to support from/to either.

@jianjuns I don't understand the comment. Without this, we cannot enforce security rules for peers (to/from) which are external endpoints. Isn't this part the responsibility of the agent? For appliedTo, there are multiple ways of enforcing rules applied to external endpoints (security groups, hypervisor, agent, etc.). So I would consider the 2 to be separate work items.

Reply:
That being said, I agree that we should postpone merging this to after the v0.9.0 release.

Reply:
ok.. i see Jianjun raised two points.. i too think we could separate the externEntities support in appliedTo and in to/from in two separate discussions?

Reply:
I think ExternalEntity is to support non-K8s managed Nodes/Entities. To be complete, we should support ExternalEntities in both from/to and appliedTo. If you really want we can choose to support from/to first, but not much value in my mind.

Reply:
It's just that one can be done easily by the Agent (and has to be done by the Agent with our current Controller design), while the other can be done in multiple ways

Reply:
@jianjuns  as @Dyanngg and @antoninbas mentioned, this change allows agent running to k8s worker node to enforce traffic to/from exteranlEntiteis(cloud resources). We don't have an agent that runs on externalEntity (VM etc) itself.

Reply:
Understood, but do you have use cases for that already?

Reply:
yes. 
use case is that  Pod gains access to services on  cloud VMs. This was part of antrea+ demo.

Reply:
Ok. Then for Antrea perspective, it is more about supporting externalEntity in from/to. Still do not feel it is very useful alone. If am ok to make it the first step, but we should either have a way to validate externalEntity cannot be in appliedTo, or have a featureGate to disable the whole externalEntity support?
Let us talk offline.

Reply:
Hi @jianjuns 

@Dyanngg has kindly taking over antrea externalEntity support. If u have any concrete suggestions, he will help out.

Regarding to ur suggestion about " ... either have a way to validate externalEntity cannot be in appliedTo, or ...", if I recall correctly, if user specifies ExternalEntity in appliedTo field of an ANP
1) controller will not forward the AppliedToGroup (contain externalEntity) to agent
2) even if controller (via bug) forward AppliedToGroup( contain externalEntity) to agent, agent ignores externalEntity in AppliedToGroup.

I hope this satisfies ur requirements.

Reply:
I would say it is very confusing to users. We should document we do not support appliedTo and validate it, or disable the externalEntity feature until we really support agent on external Nodes.

Reply:
I was chatting with @abhiraut on this. We also feel namespaced externalEntity does not make much sense from Antrea perspective. But anyway we might start from what we have now, but at least we should add some validation no externalEntity in appliedTo.